R/appshot.R

```diff
@@ -40,14 +42,59 @@ appshot.character <- function(app, file = "webshot.png", ...,
     p <- processx::process$new("R", args = c("--slave", "-e", cmd))
   })
 
+  # Make sure app is killed on exit
   on.exit({
     p$kill()
   })
 
   # Wait for app to start
   Sys.sleep(0.5)
 
+  # Get screenshot
   fileout <- webshot(sprintf("http://127.0.0.1:%d/", port), file = file, ...)
 
   invisible(fileout)
 }
+
+
+#' @rdname appshot
+#' @export
+appshot.shiny.appobj <- function(app, file = "webshot.png", ...,
+                              port = getOption("shiny.port"), envvars = NULL) {
+
+
+  port <- available_port(port)
+
+  args <- list(
+    url = sprintf("http://127.0.0.1:%d/", port),
+    file = file,
+    ...
+  )
+  r_session <- callr::r_bg(
+    function(...) {
+      # Wait for app to start
+      Sys.sleep(0.5)
+      webshot::webshot(...)
+    },
+    args
+  )
+
+  # Add a shiny app observer which checks every 200ms to see if the background r session is alive
+  shiny::observe({
+    # check the r session rather than the file to avoid race cases or random issues
+    if (r_session$is_alive()) {
+      # try again later
+      shiny::invalidateLater(200)
+    } else {
```

Comment:
It would also be good to have a timeout check, just in case the R process with webshot hangs for some reason.